source/azure_iot_hub_client.c

```diff
@@ -626,3 +724,503 @@ AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClient_t *
+                                                       ulTimeoutMilliseconds ) ) != eAzureIoTMQTTSuccess )
+    {
+        AZLogError( ( "AzureIoTMQTT_ProcessLoop failed : ProcessLoopDuration=%u, Error=0x%08x",
+                      ulTimeoutMilliseconds, xMQTTResult ) );
+        xReturn = eAzureIoTHubClientFailed;
+    }
+    else
+    {
+        xReturn = eAzureIoTHubClientSuccess;
+    }
+
+    return xReturn;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_SubscribeCloudToDeviceMessage( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                                           AzureIoTHubClientCloudToDeviceMessageCallback_t xCallback,
+                                                                           void * prvCallbackContext,
+                                                                           uint32_t ulTimeoutMilliseconds )
+{
+    AzureIoTMQTTSubscribeInfo_t xMqttSubscription = { 0 };
+    AzureIoTMQTTResult_t xMQTTResult;
+    AzureIoTHubClientResult_t xReturn;
+    uint16_t usSubscribePacketIdentifier;
+    AzureIoTHubClientReceiveContext_t * pxContext;
+
+    if( ( pxAzureIoTHubClient == NULL ) ||
+        ( xCallback == NULL ) )
+    {
+        AZLogError( ( "Failed to subscribe cloud message : Invalid argument" ) );
+        xReturn = eAzureIoTHubClientInvalidArgument;
+    }
+    else
+    {
+        pxContext = &pxAzureIoTHubClient->_internal.xReceiveContext[ azureiothubRECEIVE_CONTEXT_INDEX_C2D ];
+        xMqttSubscription.xQoS = eAzureIoTMQTTQoS1;
+        xMqttSubscription.pcTopicFilter = ( const uint8_t * ) AZ_IOT_HUB_CLIENT_C2D_SUBSCRIBE_TOPIC;
+        xMqttSubscription.usTopicFilterLength = ( uint16_t ) sizeof( AZ_IOT_HUB_CLIENT_C2D_SUBSCRIBE_TOPIC ) - 1;
+        usSubscribePacketIdentifier = AzureIoTMQTT_GetPacketId( &( pxAzureIoTHubClient->_internal.xMQTTContext ) );
+
+        AZLogDebug( ( "Attemptting to subscribe to the MQTT topic: %s.", AZ_IOT_HUB_CLIENT_C2D_SUBSCRIBE_TOPIC ) );
+
+        if( ( xMQTTResult = AzureIoTMQTT_Subscribe( &( pxAzureIoTHubClient->_internal.xMQTTContext ),
+                                                    &xMqttSubscription, 1, usSubscribePacketIdentifier ) ) != eAzureIoTMQTTSuccess )
+        {
+            AZLogError( ( "Failed to subscribe, error=0x%08x", xMQTTResult ) );
+            xReturn = eAzureIoTHubClientFailed;
+        }
+        else
+        {
+            pxContext->_internal.usState = azureiothubTOPIC_SUBSCRIBE_STATE_SUB;
+            pxContext->_internal.usMqttSubPacketID = usSubscribePacketIdentifier;
+            pxContext->_internal.pxProcessFunction = prvAzureIoTHubClientC2DProcess;
+            pxContext->_internal.callbacks.xCloudToDeviceMessageCallback = xCallback;
+            pxContext->_internal.pvCallbackContext = prvCallbackContext;
+            xReturn = prvWaitForSubAck( pxAzureIoTHubClient, pxContext, ulTimeoutMilliseconds );
```

Comment:
Should we be filling out the fields above before `prvWaitForSuback` returns successfully.  I worry there's a race where if doing ProcessLoop() in the wait, we get the SUBSCRIBE and also a PUBLISH to something app has SUB'd for in same packet then we could potentially end up invoking the application's callback *before* we returned from `AzureIoTHubClient_SubscribeCloudToDeviceMessage`.

I also would worry that if `xReturn` is a failure, we should leave the fields and callback null'd out.

Reply:
Ack

---
In reply to: [612753409](https://github.com/Azure/azure-iot-middleware-freertos/pull/75#discussion_r612753409) [](ancestors = 612753409)

Reply:
@hihigupt - tagging because you'd resolved this discussion.  I literally woke up at 3AM last night and realized there's a problem with what I've proposed, too.  Suppose we set the callback pointers after `prvWaitForSubAck` and case where the PUBLISH comes right away, before we return.  I don't think this is outside the realm of possibility with say a C2D message that the Hub had queued to send.

In this case, if we don't have callback filled, then we end up with the case where the PUBLISH to us is dropped.  Which seems worse to me than case where we do the app's callback from within the subscribe function.

I don't have much appetite for complicated state machines where we mark the message wants to be PUBLISH'd to app, but don't send it to app here.  Beyond complexity I don't know how to do that without making a copy of the message.

So I think that we should just document clearly that there are cases where the callback may be called before we return from this function and that the application should be prepared for this case.

Reply:
Great point ... Setting up of receive path before we do Process Loop is important as we might have windows to drop packets. I will set the receive path and cleanup if failed